tests/test_preview.py

```diff
@@ -0,0 +1,706 @@
+    exp.generate(db, overwrite=True)
+    rs1 = exp.create_run_settings("echo", ["hello", "world"])
+    # Create ensemble
+    ensemble = exp.create_ensemble("fd_simulation", run_settings=rs1, replicas=2)
+    # enable key prefixing on ensemble
+    ensemble.enable_key_prefixing()
+    exp.generate(ensemble, overwrite=True)
+    rs2 = exp.create_run_settings("echo", ["spam", "eggs"])
+    # Create model
+    ml_model = exp.create_model("tf_training", rs2)
+
+    for sim in ensemble.entities:
+        ml_model.register_incoming_entity(sim)
+
+    exp.generate(ml_model, overwrite=True)
+    preview_manifest = Manifest(db, ml_model, ensemble)
+
+    # Call preview renderer for testing output
+    output = previewrenderer.render(exp, preview_manifest)
+
+    assert "Client Configuration" in output
+    assert "Database identifier" in output
+    assert "Database backend" in output
+    assert "Type" in output
+
+
+def test_ensemble_preview_attached_files(fileutils, test_dir, wlmutils):
+    """
+    Test the preview of tagged, copy, and symlink files attached
+    to an ensemble
+    """
+    # Prepare entities
+    test_launcher = wlmutils.get_test_launcher()
+    exp = Experiment("attached-files-test", exp_path=test_dir, launcher=test_launcher)
+    ensemble = exp.create_ensemble(
+        "dir_test", replicas=1, run_settings=RunSettings("python", exe_args="sleep.py")
+    )
+    ensemble.entities = []
+    params = {"THERMO": [10, 20], "STEPS": [20, 30]}
+    ensemble = exp.create_ensemble(
+        "dir_test",
+        params=params,
+        run_settings=RunSettings("python", exe_args="sleep.py"),
+    )
+    gen_dir = fileutils.get_test_conf_path(osp.join("generator_files", "test_dir"))
+    symlink_dir = fileutils.get_test_conf_path(
+        osp.join("generator_files", "to_symlink_dir")
+    )
+    copy_dir = fileutils.get_test_conf_path(osp.join("generator_files", "to_copy_dir"))
+
+    ensemble.attach_generator_files()
+    ensemble.attach_generator_files(
+        to_configure=[gen_dir, copy_dir], to_copy=copy_dir, to_symlink=symlink_dir
+    )
+    preview_manifest = Manifest(ensemble)
+
+    # Call preview renderer for testing output
+    output = previewrenderer.render(exp, preview_manifest)
+
+    # Evaluate output
+    assert "Tagged Files for model configuration" in output
+    assert "Copy files" in output
+    assert "Symlink" in output
+    assert "Ensemble parameters" in output
+    assert "Model parameters" in output
+
+    assert "generator_files/test_dir" in output
+    assert "generator_files/to_copy_dir" in output
+    assert "generator_files/to_symlink_dir" in output
+    assert "/SmartSim" in output
+    for model in ensemble:
+        assert "generator_files/test_dir" in model.files.tagged[0]
+        for copy in model.files.copy:
+            assert "generator_files/to_copy_dir" in copy
+        for link in model.files.link:
+            assert "generator_files/to_symlink_dir" in link
```

Comment:
Does each copy/link in each ensemble always get the same value? if they have slight variations

- /smartsim/generator/test/*modela*/generator_files/to_copy
- /smartsim/generator/test/*modelb*/generator_files/to_copy
- /smartsim/generator/test/*modelc*/generator_files/to_copy

it might be worthwhile to ensure a bug didn't crop up where we put the same thing for all of them through a template mistake. If they are the same, we're good.

Reply:
paths all look the same.